---
title: "homomine report"
author: "Sanzhen Liu"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  cwd: NUL
  scriptdir: NULL
  qrygene: NULL
  qryseq: NULL
  qrybase: NULL
  qrybed: NULL
  tgtbase: NULL
  tgtgene: NULL
  tgtseq: NULL
  tgtbed: NULL
  tgthit: NULL
  datadir: NULL
  figuredir: NULL
  qryinput: NULL
  nucmeraln: NULL
output:
  html_document: default
  word_document: default
  pdf_document: default
---

```{r parameters, echo = FALSE}
library(knitr)
qrygene <- params$qrygene
cwd <- params$cwd
scriptdir <- params$scriptdir
qrygene <- params$qrygene
qrybase <- params$qrybase
qryseq <- params$qryseq
qrybed <- params$qrybed
tgtgene <- params$tgtgene
tgtbase <- params$tgtbase
tgthit <- params$tgthit
tgtseq <- params$tgtseq
tgtbed <- params$tgtbed
tgthit <- params$tgthit
datadir <- params$datadir
figuredir <- params$figuredir
qryinput <- params$qryinput
nucmeraln <- params$nucmeraln
```

### Homomine report of **`r qrygene`**

#### o Basic information of `r qrygene` from `r qrybase`
```{r qry, echo=F, comment=""}
qry_data <- read.delim(qryinput)
kable(qry_data)
```

#### o Target hit in `r tgtbase`
```{r target, echo=F, comment=""}
tgt_data <- read.delim(tgthit)
kable(tgt_data)
```

```{r match, echo=F, comment=""}
qt_data <- rbind(qry_data[, 1:5], tgt_data[, 1:5])
qt_data$genome <- c(qrybase, tgtbase)
qt_data_outfile <- paste0(cwd, "/", qrygene, "/", qrygene, "_", tgtbase, ".homo.txt")
write.table(qt_data, qt_data_outfile, row.names=F, quote=F, sep="\t")
```

#### o Virsualization of Nucmer alignments
```{r comp, echo=F, comment="", fig.height=3, fig.width=6, fig.cap="Fig 1. Nucmer alignment."}
source(paste0(scriptdir, "/homomine_nucmerplot.R"))
outpdf <- paste0(qrygene, "_to_", tgtbase, "nucmer.pdf")
homomine_nucmerplot(datafile=nucmeraln, qrybed=qrybed, tgtbed=tgtbed,
                    outpath=figuredir, ispdf=F, imageoutfile=outpdf)
```
Nucmer was used to align sequences of the two sequences. Only uniquely aigned regions are displayed. Green: repetitive sequences; light blue: untranslated regions; blue: exons.

```{r aln, echo=F, comment=""}
aln_data <- read.delim(nucmeraln)
kable(aln_data)
```

#### o Genomic variants with high or moderate impacts
```{r vcfeff, echo=F, comment="", results = 'asis'}
vcfeff_files <- dir(path=datadir, pattern="eff$", full.names=T)
if (length(vcfeff_files)>0) {
for (evcfeff in vcfeff_files) {
  evcfeff_data <- read.delim(evcfeff, check.names=F)
  print(kable(evcfeff_data))
  cat("\n")
}
} else {
  cat("    No high or moderate variants identified\n")
}
```

#### o All polymorphisms identified
```{r vcf, echo=F, comment="", results = 'asis'}
vcffiles <- dir(path=datadir, pattern="vcf$", full.names=T)
max_seqlen2display <- 50
variant_found <- 0
if (length(vcffiles) > 0) {
  for (evcf in vcffiles) {
    evcf_data <- read.delim(evcf, comment.char="#", header=F)
    if (nrow(evcf_data) > 0) {
      variant_found <- 1
    }
    evcf_data <- evcf_data[, c(1,2,4,5)]
    colnames(evcf_data) <- c("gene", "pos", "REF", "ALT")
    ref_long <- nchar(evcf_data$REF) > max_seqlen2display
    alt_long <- nchar(evcf_data$ALT) > max_seqlen2display
    evcf_data$REF[ref_long] <- paste0("long-", nchar(evcf_data$REF)[ref_long], "bp")
    evcf_data$ALT[alt_long] <- paste0("long-", nchar(evcf_data$ALT)[ref_long], "bp")
    print(kable(evcf_data))
  }
} else {
  cat("    No variant files found\n")
}

if (variant_found == 0) {
  cat("    No variants found\n")
}
```

#### o Query sequence from `r qrybase`
```{r qryseq, echo=F, comment=""}
qry_seq <- readLines(qryseq)
qry_seqname <- qry_seq[1]
qry_seqdata <- paste(qry_seq[-1], collapse = "")
cat(qry_seqname, qry_seqdata, sep="\n")
cat("\n")
```

#### o Target sequence from `r tgtbase`
```{r tgtseq, echo=F, comment=""}
tgt_seq <- readLines(tgtseq)
tgt_seqname <- tgt_seq[1]
tgt_seqdata <- paste(tgt_seq[-1], collapse = "")
cat(tgt_seqname, tgt_seqdata, sep="\n")
cat("\n")
```