---
title: "homomine report"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  cwd: NUL
  scriptdir: NULL
  qrygene: NULL
  qryseq: NULL
  qrybase: NULL
  qrybed: NULL
  tgtbase: NULL
  tgtgene: NULL
  tgtseq: NULL
  tgtbed: NULL
  tgthit: NULL
  qrysvte: NULL
  tgtsvte: NULL
  datadir: NULL
  figuredir: NULL
  qryinput: NULL
  nucmeraln: NULL
output:
  html_document: default
  word_document: default
  pdf_document: default
---

```{r parameters, echo = FALSE}
library(knitr)
qrygene <- params$qrygene
cwd <- params$cwd
scriptdir <- params$scriptdir
qrygene <- params$qrygene
qrybase <- params$qrybase
qryseq <- params$qryseq
qrybed <- params$qrybed
tgtgene <- params$tgtgene
tgtbase <- params$tgtbase
tgthit <- params$tgthit
tgtseq <- params$tgtseq
tgtbed <- params$tgtbed
tgthit <- params$tgthit
qrysvte <- params$qrysvte
tgtsvte <- params$tgtsvte
datadir <- params$datadir
figuredir <- params$figuredir
qryinput <- params$qryinput
nucmeraln <- params$nucmeraln
```

## Homomine report of **`r qrygene`**

### o Basic information of `r qrygene` from `r qrybase`
```{r qry, echo=F, comment=""}
if (file.exists(tgthit)) {
  qry_data <- read.delim(qryinput)
  kable(qry_data)
} else {
  cat(paste(tgthit, " does not exist\n"))
  knitr::knit_exit()
}
```

### o Target hit in `r tgtbase`
*Note*: Below chromosomal interval might not cover the whole gene.  
```{r target, echo=F, comment=""}
if (file.exists(tgthit)) {
  tgt_data <- read.delim(tgthit)
  kable(tgt_data)
} else {
  cat(paste(tgthit, " does not exist\n"))
  knit_exit()
}
```

```{r match, echo=F, comment=""}
qt_data <- rbind(qry_data[, 1:5], tgt_data[, 1:5])
qt_data$genome <- c(qrybase, tgtbase)
qt_data_outfile <- paste0(datadir, "/../../", qrygene, "_", tgtbase, ".homo.txt")
write.table(qt_data, qt_data_outfile, row.names=F, quote=F, sep="\t")
```

### o Alignments between the two sequences
```{r alncheck, echo=F, comment=""}
# check alignment
aln_check <- read.delim(nucmeraln, header=F)
if (nrow(aln_check)==1) {
  cat(" no alignments are in the Nucmer output (uniqure and filted output)\n")
  knit_exit()
}
```

```{r comp, echo=F, comment="", fig.height=3, fig.width=6, fig.cap="**Fig 1. Visualization of alignments between the two sequences**"}
source(paste0(scriptdir, "/homomine_nucmerplot.R"))
# plotting
outpdf <- paste0(qrygene, "_to_", tgtbase, "nucmer.pdf")
homomine_nucmerplot(datafile=nucmeraln, qrybed=qrybed, tgtbed=tgtbed,
                    outpath=figuredir, ispdf=T, imageoutfile=outpdf)

homomine_nucmerplot(datafile=nucmeraln, qrybed=qrybed, tgtbed=tgtbed,
                    outpath=figuredir, ispdf=F, imageoutfile=outpdf)
```
NUCMER was used to align sequences of the two sequences. Only uniquely aigned regions are displayed. Orange: repetitive sequences; light blue: untranslated exon regions; blue: coding regions.

```{r aln, echo=F, comment=""}
aln_data <- read.delim(nucmeraln)
kable(aln_data, caption="<strong>Table. Details of sequence alignments</strong>")
```
*Notes*:  
qstart, qend, qmatch, qlen, qry represent start, end, matched length, total length, and query name in `r qrybase`, respectively.  
sstart, send, smatch, slen, subj represent start, end, matched length, total length, and target name in `r tgtbase`, respectively.  
identity: percentage of the alignment identity (%)  

### o Genomic variants with high or moderate impacts
```{r vcfeff, echo=F, comment="", results='asis'}
vcfeff_files <- dir(path=datadir, pattern="eff$", full.names=T)
if (length(vcfeff_files)>0) {
  transcript_count <- 0
  for (evcfeff in vcfeff_files) {
    transcript_count <- transcript_count + 1
    evcfeff_data <- read.delim(evcfeff, check.names=F)
    if (nrow(evcfeff_data) > 0) {
      tablehead <- paste0("<strong>Table. Variants with a high or moderate impact in transcript ", transcript_count, "</strong>")
	  print(kable(evcfeff_data, , caption=tablehead))
      cat("\n")
    } else {
	  cat("--------------------------------------------------------\n")
      cat("<strong>No high or moderate variants identified for transcript ", transcript_count, "</strong>\n", sep="")
	}
  }
} else {
  cat("No high or moderate variants identified\n")
}
```

### o Structural variation used `r qrybase` as the reference
SV was defined if either REF or ALT allele has at least 50 bp and the length difference of two alleles are at least 30 bp.  
```{r qrySV, echo=F, comment="", results='asis'}
if (file.info(qrysvte)$size > 0) {
  qrysvte_data <- read.delim(qrysvte, header=F)
  if (ncol(qrysvte_data)<=5) {
    colnames(qrysvte_data) <- c("seqname", "start", "end", "svtype", "svnote")
  } else {
    colnames(qrysvte_data) <- c("seqname", "start", "end", "svtype", "svnote", "TEstart", "TEend", "TE")
  }
  
  if (nrow(qrysvte_data)>0) {
    qrysvte_data$start <- qrysvte_data$start + 1 # adjust to 1-based
    if (ncol(qrysvte_data)>5) {
		qrysvte_data$TEstart[qrysvte_data$TEstart >= 0] <- qrysvte_data$TEstart[qrysvte_data$TEstart >= 0] + 1
	}
	print(kable(qrysvte_data, caption="**Table. SV and potential TEs**"))
  	qrysvte_outfile <- paste0(datadir, "/../../", qrygene, "_", tgtbase, ".SV.txt")
	write.table(qrysvte_data, qrysvte_outfile, row.names=F, quote=F, sep="\t")
  } else {
    cat("No SVs identified\n")
  }
} else {
  cat("No SVs identified\n")
}
```

### o Structural variation used `r tgtbase` as the reference
```{r tgtSV, echo=F, comment="", results='asis'}
if (file.info(tgtsvte)$size > 0) {
  tgtsvte_data <- read.delim(tgtsvte, header=F)
  if (ncol(tgtsvte_data)<=5) {
    colnames(tgtsvte_data) <- c("seqname", "start", "end", "svtype", "svnote")
  } else {
    colnames(tgtsvte_data) <- c("seqname", "start", "end", "svtype", "svnote", "TEstart", "TEend", "TE")
  }
  
  if (nrow(tgtsvte_data)>0) {
	tgtsvte_data$start <- tgtsvte_data$start + 1 # adjust to 1-based
	if (ncol(tgtsvte_data)>5) {
		tgtsvte_data$TEstart[tgtsvte_data$TEstart >= 0] <- tgtsvte_data$TEstart[tgtsvte_data$TEstart >= 0] + 1
	}
	print(kable(tgtsvte_data, caption="**Table. SV and potential TEs**"))
	tgtsvte_outfile <- paste0(datadir, "/../../", tgtbase, "_", qrygene, ".SV.txt")
	write.table(tgtsvte_data, tgtsvte_outfile, row.names=F, quote=F, sep="\t")
  } else {
    cat("No SVs identified\n")
  }
} else {
  cat("No SVs identified\n")
}
```

### o All polymorphisms identified
```{r vcf, echo=F, comment="", results='asis'}
vcffiles <- dir(path=datadir, pattern="vcf$", full.names=T)
max_seqlen2display <- 50
variant_found <- 0
if (length(vcffiles) > 0) {
  for (evcf in vcffiles) {
    evcf_refname <- gsub(".*vars_on_", "", evcf)
    evcf_refname <- gsub(".vcf$", "", evcf_refname)
    #evcf_data <- read.delim(evcf, comment.char="#", header=F)
    evcf_data <- try(read.delim(evcf, comment.char="#", header=F, colClasses="character"), silent=T)
    if (!inherits(evcf_data, 'try-error')) {
      variant_found <- 1
      evcf_data <- evcf_data[, c(1,2,4,5)]
      colnames(evcf_data) <- c("seqname", "pos", "REF", "ALT")
      ref_long <- nchar(evcf_data$REF) > max_seqlen2display
      alt_long <- nchar(evcf_data$ALT) > max_seqlen2display
      evcf_data$REF[ref_long] <- paste0("long-", nchar(evcf_data$REF)[ref_long], "bp")
      evcf_data$ALT[alt_long] <- paste0("long-", nchar(evcf_data$ALT)[alt_long], "bp")
      print(kable(evcf_data,
	              caption=paste0("<strong>Table. Variants in the sequence from ", evcf_refname, "</strong>")))
    }
  }
} else {
  cat("    No variant files found\n")
}

if (variant_found == 0) {
  cat("    No variants found\n")
}
```

### o Query sequence from `r qrybase`
```{r qryseq, echo=F, comment=""}
qry_seq <- readLines(qryseq)
qry_seqname <- qry_seq[1]
qry_seqdata <- paste(qry_seq[-1], collapse = "")
cat(qry_seqname, qry_seqdata, sep="\n")
cat("\n")
```

### o Target sequence from `r tgtbase`
```{r tgtseq, echo=F, comment=""}
tgt_seq <- readLines(tgtseq)
tgt_seqname <- tgt_seq[1]
tgt_seqdata <- paste(tgt_seq[-1], collapse = "")
cat(tgt_seqname, tgt_seqdata, sep="\n")
cat("\n")
```
